server/ServerApp.cpp

```diff
@@ -965,6 +966,52 @@ namespace {
                                    << "human player with assigned empire id: " << psd.m_save_game_empire_id;
         }
     }
+
+    /** Replace species selection for any players set to "RANDOM" species, with a species based on @p initial_seed and player id.
+     *  If changed, seed is reset to @p initial_seed before return. */
+    void ResolveRandomSpeciesSelections(std::map<int, PlayerSetupData>& player_setup_data, const unsigned int initial_seed) {
+        SpeciesManager& sm = GetSpeciesManager();
+        int num_playable_species = sm.NumPlayableSpecies();
+        // early return for no playable species
+        if (num_playable_species <= 0) {
+            ErrorLogger() << "No playable species found";
+            return;
+        }
+        bool seed_altered = false;
+
+        // evenly initialize a list for weighting each species
+        float prob_init_value = 1.0f / num_playable_species;
+        std::vector<float> species_prob;
+        for (int i = 0; i < num_playable_species; ++i)
+            species_prob.push_back(prob_init_value);
+
+        float weight_reduction = prob_init_value / num_playable_species;
+
+        // iterate over players
+        for (std::map<int, PlayerSetupData>::value_type& entry : player_setup_data) {
+            if (entry.second.m_starting_species_name == "RANDOM") {
+                int player_id = entry.first;
+                // alter the seed for each player according to their id
+                seed_altered = true;
+                Seed(initial_seed + player_id);
+                unsigned int seed = initial_seed * RandGaussian(1.0, (player_id * 0.01));
+                Seed(seed);
+                int idx = RandWeightedInt(species_prob);
+
+                // reduce the chance of this species being picked again
+                species_prob[idx] -= weight_reduction;
```

Comment:
Thank you for this.  Even for relative weights, it would be better not to tempt an issue with negative values.  I will keep this in mind if I implement anything similar in the future.